File: medium/34. Find First and Last Position of Element in Sorted Array/34. Find First and Last Position of Element in Sorted Array.ts
function searchRange(nums: number[], target: number): number[] {
    const first = binarySearch(nums, target, true)
    const second = binarySearch(nums, target, false)
    return [first, second]
};


const binarySearch = (nums: number[], target: number, isFirst: boolean): number => {
    let left = 0, right = nums.length - 1, pos = -1

    while (left <= right) {
        const middle = left + Math.floor((right - left) / 2)
        if (nums[middle] === target) {
            pos = middle
            if (isFirst) {
                right = middle - 1
            } else {
                left = middle + 1
            }
        } else if (nums[middle] < target) {
            left = middle + 1
        } else {
            right = middle - 1
        }
    }

    return pos
}


console.log(searchRange([5, 7, 7, 8, 8, 10], 8))
console.log(searchRange([5, 7, 7, 8, 8, 10], 6))
console.log(searchRange([8, 8], 8))
console.log(searchRange([1], 1))

export { }

File: medium/34. Find First and Last Position of Element in Sorted Array/34. Find First and Last Position of Element in Sorted Array.py
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        # left, right = 0, len(nums) - 1

        # position = [-1, -1]

        # while left <= right:
        #     middle = (left + right) // 2

        #     if nums[middle] == target:
        #         while middle and nums[middle - 1] == target:
        #             middle -= 1

        #         position[0] = middle

        #         while middle <= right and nums[middle] == target:
        #             middle += 1

        #         position[1] = middle - 1
        #         return position

        #     if nums[left] <= nums[middle]:
        #         if nums[left] <= target < nums[middle]:
        #             right = middle - 1
        #         else:
        #             left = middle + 1
        #     else:
        #         if nums[middle] < target <= nums[right]:
        #             left = middle + 1
        #         else:
        #             right = middle - 1

        # return position

        first = self.binarySearch(nums, target, True)
        second = self.binarySearch(nums, target, False)

        return [first, second]

    def binarySearch(self, nums: List[int], target: int, isFirst: bool) -> int:
        low, high = 0, len(nums) - 1
        pos = -1

        while low <= high:
            middle = low + (high - low) // 2
            if nums[middle] == target:
                pos = middle
                if isFirst:
                    high = middle - 1
                else:
                    low = middle + 1
            elif nums[middle] < target:
                low = middle + 1
            else:
                high = middle - 1
        return pos


solution = Solution()
print(solution.searchRange(nums=[5, 7, 7, 8, 8, 10], target=8))
print(solution.searchRange(nums=[5, 7, 7, 8, 8, 10], target=6))
print(solution.searchRange(nums=[8, 8], target=8))
print(solution.searchRange(nums=[1], target=1))
